### PR TITLE
Add declaration-expressions.md proposal

### PR DESCRIPTION
**Pull request by cston**
Direct link to readable doc: https://github.com/dotnet/csharplang/pull/977/files?short_path=9ee3b9d#diff-9ee3b9d686b057ca42683cf843811bc5

### CODE REVIEW

**Comment by cston** [comment]
cc @gafter, @mattwar

File: proposals/declaration-expressions.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Declaration expressions
```
Comment (gafter) [inline proposals/declaration-expressions.md]: This section at the top is not needed in the proposal. It is used in the championed issue referring to the proposal.

File: proposals/declaration-expressions.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    ;
+declaration_expression // C# 7.0
+    | type variable_designation
+    ;
+```
+
+The declaration assignment is of a single local.
+```c#
+F(var x, y = 2);        // error
+F(var (x, y) = (1, 2)); // error
```
Comment (gafter) [inline proposals/declaration-expressions.md]: This form is likely to be supported by our relaxation of deconstruction, a separate feature.

File: proposals/declaration-expressions.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+The declaration assignment expression may be an l-value, for `ref` argument values in particular.
+
+If the declaration assignment expression declares a value type, and the expression is an r-value, the value of
+the expression is a copy.
+
+The scope of locals declared in declaration assignment expressions is the same the scope of corresponding declaration expressions from C#7.0.
+
+It is a compile time error to refer to a local in text preceding the declaration expression.
+
```
Comment (gafter) [inline proposals/declaration-expressions.md]: I think we should say something about whether ref locals are permitted.

File: proposals/declaration-expressions.md (hunk 2 above)
Comment (cston) [reply proposals/declaration-expressions.md]: Removed.

File: proposals/declaration-expressions.md (hunk 1 above)
Comment (cston) [reply proposals/declaration-expressions.md]: Removed.

File: proposals/declaration-expressions.md (hunk 3 above)
Comment (cston) [reply proposals/declaration-expressions.md]: Added.

**Comment by cston** [comment]
Adding @dotnet/roslyn-compiler.

**Comment by cston** [comment]
Adding @MadsTorgersen.

**Pull request merged**